python/lbann/modules/base.py

```diff
@@ -0,0 +1,233 @@
+    global_count = 0  # Static counter, used for default names
+
+    def __init__(self, size, bias=True, weights=[], activation=None,
+                 name=None, data_layout='data_parallel'):
+        """Initialize fully-connected module.
+
+        Args:
+            size (int): Size of output tensor.
+            activation (type): Layer class for activation function.
+            bias (bool): Whether to apply bias after linearity.
+            weights (`Weights` or iterator of `Weights`): Weights in
+                fully-connected layer. There are at most two: the
+                matrix and the bias. If weights are not provided, the
+                matrix will be initialized with He normal
+                initialization and the bias with zeros.
+            name (str): Default name is in the form 'fcmodule<index>'.
+            data_layout (str): Data layout.
+
+        """
+        super().__init__()
+        FullyConnectedModule.global_count += 1
+        self.instance = 0
+        self.size = size
+        self.bias = bias
+        self.name = (name
+                     if name
+                     else 'fcmodule{0}'.format(FullyConnectedModule.global_count))
+        self.data_layout = data_layout
+
+        # Initialize weights
+        # Note: If weights are not provided, matrix weights are
+        # initialized with He normal scheme and bias weights are
+        # initialized with zeros.
+        self.weights = list(make_iterable(weights))
+        if len(self.weights) > 2:
+            raise ValueError('`FullyConnectedModule` has '
+                             'at most two weights, '
+                             'but got {0}'.format(len(self.weights)))
+        if len(self.weights) == 0:
+            self.weights.append(
+                lbann.Weights(initializer=lbann.HeNormalInitializer(),
+                              name=self.name+'_matrix'))
+        if len(self.weights) == 1:
+            self.weights.append(
+                lbann.Weights(initializer=lbann.ConstantInitializer(value=0.0),
+                              name=self.name+'_bias'))
+
+        # Initialize activation layer
+        self.activation = None
+        if activation:
+            if isinstance(activation, type):
+                self.activation = activation
+            else:
+                self.activation = type(activation)
+            if not issubclass(self.activation, lbann.Layer):
+                raise ValueError('activation must be a layer')
+
+    def forward(self, x):
+        self.instance += 1
+        name = '{0}_instance{1}'.format(self.name, self.instance)
+        y = lbann.FullyConnected(x,
+                                 weights=self.weights,
+                                 name=(name+'_fc' if self.activation else name),
+                                 data_layout=self.data_layout,
+                                 num_neurons=self.size,
+                                 has_bias=self.bias)
+        if self.activation:
```

Comment:
How do you deal with activations with additional parameters e.g., LeakyRelu has slope?

Reply:
I'm leaning against supporting this. [`torch.nn.modules.Linear`](https://pytorch.org/docs/stable/_modules/torch/nn/modules/linear.html) doesn't support activations; [`tf.keras.layers.Dense`](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dense) and [`keras.layers.Dense`](https://keras.io/layers/core/) don't allow you to pass arguments to the activation. We might be able to make this usable for some cases by supporting non-zero defaults for leaky reLU and ELU.